brickstrap-image.sh

```diff
@@ -0,0 +1,221 @@
+    if [ -z "$(brp_get_image_type_driver "$1")" ]; then
+        eval "BRP_IMG_DRV_REGISTRY_$1=$2"
+        eval "BRP_IMG_EXT_REGISTRY_$1=$3"
+        if [ -n "$4" ]; then
+            eval "BRP_IMG_CFG_VALIDATOR_$1=$4"
+        fi
+    else
+        fail "Rejected duplicate driver '$2' for image type: '$1'
+Previous setting was: $(brp_get_image_type_driver "$1")"
+    fi
+}
+
+#
+# Validates image driver settings in the project configuration for the given
+# drivers (names). This function works similar to brp_create_image_types()
+# $1: list of drivers for which any configuration validators should be invoked
+# $2: list of previously checked driver configurations (if applicable).
+#
+function brp_validate_img_configs()
+{
+    [ $# -eq 1 -a -n "$1" ] && for BRP_CUR_IMG in $1; do
+        eval "BRP_CUR_IMG=$BRP_CUR_IMG" # unwrap quotes
+        if [ -n "$(brp_get_image_cfg_validator "$BRP_CUR_IMG")" ]; then
+            echo "$2" | fgrep -q "'$BRP_CUR_IMG'" || \
+                eval "$(brp_get_image_cfg_validator "$BRP_CUR_IMG")"
+        fi
+    done
+}
+
+#
+# Validates image driver specific settings in the project configuration.
+#
+function brp_validate_image_configurations()
+{
+    if [ -n "$BR_IMAGE_TYPES" ]; then
+        brp_validate_img_configs "$BR_IMAGE_TYPES"
+    fi
+    if [ -n "$IMAGE_TYPES" ]; then
+        brp_validate_img_configs "$IMAGE_TYPES" "$BR_IMAGE_TYPES"
+    fi
+    if [ -z "$IMAGE_TYPES" -a -z "$BR_IMAGE_TYPES" ]; then
+        brp_validate_default_image_config
+    fi
+}
+
+#
+# Creates disk image files for a given list of image types to create.
+# $1: list of image types for which an image file should be created
+# $2: list of previously created images (if applicable). This is used to avoid
+#    attempting to generate the same image twice.
+#
+function brp_create_image_types()
+{
+    BRP_IMAGE_TYPE_RETURN_CODE=0
+    [ $# -eq 1 -a -n "$1" ] && for BRP_CUR_IMG in "$1"; do
+        eval BRP_CUR_IMG="$BRP_CUR_IMG"
+        echo "$2" | fgrep -q "'$BRP_CUR_IMG'" || \
+            brp_create_image_type "$BRP_CUR_IMG" || \
+                BRP_IMAGE_TYPE_RETURN_CODE=$?
+    done && return $BRP_IMAGE_TYPE_RETURN_CODE
+}
+
+#
+# Creates disk images for the configured image types.
+#
+function brp_create_images()
+{
+    info "Creating image files..."
+    debug "TARBALL: $(br_tarball_path)"
+    [ ! -f "$(br_tarball_path)" ] && fail "Could not find $(br_tarball_path)"
+
+    mkdir -p "$(br_image_dir)"
+
+    BRP_IMAGE_RETURN_CODE=0
+    if [ -n "$BR_IMAGE_TYPES" ]; then
+        brp_create_image_types "$BR_IMAGE_TYPES" || BRP_IMAGE_RETURN_CODE=$?
+    fi
+    if [ -n "$IMAGE_TYPES" ]; then
+        brp_create_image_types "$IMAGE_TYPES" "$BR_IMAGE_TYPES" || \
+            BRP_IMAGE_RETURN_CODE=$?
+    fi
+    if [ -z "$BR_IMAGE_TYPES" -a -z "$IMAGE_TYPES" ]; then
+        brp_create_image_type img || BRP_IMAGE_RETURN_CODE=$?
+    fi
+
+    return $BRP_IMAGE_RETURN_CODE
```

Comment:
If you are building more than one image and the first one fails and later ones succeed, won't this function still return success? It seems like if one fails we should just be calling `fail` and stopping brickstrap altogether.